In [1]:
Output = '/Users/alexis/Library/CloudStorage/OneDrive-UniversityofNorthCarolinaatChapelHill/CEMALB_DataAnalysisPM/Projects/P1015. Fire Sufficient Similarity/3. Analyses/2. Data Processing/Output'
cur_date = '030125'

library(missForest)
library(readxl)
library(openxlsx)
library(tidyverse)
library(imputeLCMD)
library(factoextra)

# reading in files
ws_df = data.frame(Data = 'WS', read_excel("Input/Woodsmoke_Data_012825.xlsx", sheet = 2)) %>%
    select(-Sample_Number)
wf_df = data.frame(Data = 'WF', read_excel("Input/Wildfire_Data_012825.xlsx", sheet = 2)) %>%
    select(-Sample_Number)

── Attaching core tidyverse packages ──────────────────────── tidyverse 2.0.0 ──
✔ dplyr     1.1.3     ✔ readr     2.1.4
✔ forcats   1.0.0     ✔ stringr   1.5.0
✔ ggplot2   3.4.3     ✔ tibble    3.2.1
✔ lubridate 1.9.2     ✔ tidyr     1.3.0
✔ purrr     1.0.2     
── Conflicts ────────────────────────────────────────── tidyverse_conflicts() ──
✖ dplyr::filter() masks stats::filter()
✖ dplyr::lag()    masks stats::lag()
ℹ Use the conflicted package (<http://conflicted.r-lib.org/>) to force all conflicts to become errors
Loading required package: tmvtnorm

Loading required package: mvtnorm

Loading required package: Matrix


Attaching package: ‘Matrix’


The following objects are masked from ‘package:tidyr’:

    expand, pack, unpack


Loading required package: stats4

Loading required package: gmm

Loading required package: sandwich

Loading required package: norm

This package has some major limitations
(for example, it does not work reliably when
the number of variables exceeds 30),
an

NOTES TO SELF:
- change 'replicate' to 'instance' and 'chemical class' to 'class'

In [2]:
head(ws_df)
head(wf_df)

Data HAWC_ID Study                    Replicate Chemical_Class Metric
1 WS    821855 Erlandsson et al. 2020    1        PAH            Weight
2 WS    821855 Erlandsson et al. 2020    1        PAH            Weight
3 WS   1257056 McCarrick et al. 2024     1        PAH            Weight
4 WS    267140 Alfheim and Ramdahl 1984  1        PAH            Volume
5 WS   1263480 Burnet et al. 1990        1        PAH            Volume
6 WS   1263480 Burnet et al. 1990       NA        PAH            Volume
  DTXSID        Name               Value             
1 DTXSID3074787 1-Methylanthracene 11.3              
2 DTXSID3074787 1-Methylanthracene 2.2999999999999998
3 DTXSID3074787 1-Methylanthracene 1.66              
4 DTXSID3074787 1-Methylanthracene NA                
5 DTXSID3074787 1-Methylanthracene NA                
6 DTXSID3074787 1-Methylanthracene NA

Data HAWC_ID Study                Replicate Chemical_Class Metric
1 WF   1289821 Liang et al. 2021     1        PAH            Volume
2 WF   1289821 Liang et al. 2021     1        PAH            Volume
3 WF   1289737 Campbell et al. 2024  1        PAH            Weight
4 WF   1289739 Campos et al. 2019    1        PAH            Weight
5 WF   1289739 Campos et al. 2019   NA        PAH            Weight
6 WF   1289739 Campos et al. 2019   NA        PAH            Weight
  DTXSID         Name                                Value
1 NA             1-(10-methylanthracen-9-yl)ethanone 5    
2 DTXSID50176885 1-Acenaphthenone                    1    
3 DTXSID1074759  1-Methylchrysene                    NA   
4 DTXSID1074759  1-Methylchrysene                    NA   
5 DTXSID1074759  1-Methylchrysene                    NA   
6 DTXSID1074759  1-Methylchrysene                    NA

In [3]:
dim(ws_df)
dim(wf_df)

[1] 6607    9

[1] 4340    9

# Filter 1

Removing duplicate records for instances that weren't measured at all (MAR). 

In [4]:
`%notin%` <- Negate(`%in%`)

ws_df = ws_df %>%
    filter(Replicate %notin% NA)

wf_df = wf_df %>%
    filter(Replicate %notin% NA)

In [5]:
dim(ws_df)
dim(wf_df)

[1] 4203    9

[1] 2378    9

In [6]:
# combining dfs
combined_df = rbind(ws_df, wf_df)
head(combined_df)

Data HAWC_ID Study                    Replicate Chemical_Class Metric
1 WS    821855 Erlandsson et al. 2020   1         PAH            Weight
2 WS    821855 Erlandsson et al. 2020   1         PAH            Weight
3 WS   1257056 McCarrick et al. 2024    1         PAH            Weight
4 WS    267140 Alfheim and Ramdahl 1984 1         PAH            Volume
5 WS   1263480 Burnet et al. 1990       1         PAH            Volume
6 WS    267091 Forchhammer et al. 2012  1         PAH            Volume
  DTXSID        Name               Value             
1 DTXSID3074787 1-Methylanthracene 11.3              
2 DTXSID3074787 1-Methylanthracene 2.2999999999999998
3 DTXSID3074787 1-Methylanthracene 1.66              
4 DTXSID3074787 1-Methylanthracene NA                
5 DTXSID3074787 1-Methylanthracene NA                
6 DTXSID3074787 1-Methylanthracene NA

In [7]:
length(unique(combined_df$Name))

[1] 192

192 unique variables.

# Variable Background Filters

Determining what variables have missing data for all variables delineated based on the dfs the data will be split into and analyzed. 

1. The data needs to have at least two experimental (ie. measured or ND) value in both WS and WF data within each metric (ie. weight or volume) and study.
2. A variable (ie. `Name` which represents a chemical or metal) will be removed if < 50% of its experimental data points were either measurable values and/or specified as non-detects. In other words, chemicals were excluded if they were NAs (not evaluated) across <= 50% of the samples.

In [8]:
sample_type_presence_df = combined_df %>%
    # if an experimental value is there count it as being "present"
    mutate(count = ifelse(Value %notin% c('NA', 'ND'), 1, 0)) %>%
    # determining which have at least one experimental value within each sample type
    group_by(HAWC_ID, Metric, Name) %>%
    # summing the number of experimental records for each variable?????????
    reframe(Variable_Presence_Percentage = (sum(count)/n()) * 100) %>%
    filter(Variable_Presence_Percentage >= 50)

head(sample_type_presence_df)

HAWC_ID Metric Name                 Variable_Presence_Percentage
1 266980  Volume Acetaldehyde         100                         
2 266980  Volume Benzene              100                         
3 266980  Volume Benzo(GHI)perylene   100                         
4 266980  Volume Benzo(a)pyrene       100                         
5 266980  Volume Benzo(b)fluoranthene 100                         
6 266980  Volume Benzo(k)fluoranthene 100

278 data records had at least two experimental values and were retained. 

In [9]:
# filtering the original df
filter1_df = inner_join(sample_type_presence_df[,1:3], combined_df)
head(filter1_df)

Joining with `by = join_by(HAWC_ID, Metric, Name)`


HAWC_ID Metric Name               Data Study                 Replicate
1 266980  Volume Acetaldehyde       WS   Stockfelt et al. 2012 1        
2 266980  Volume Acetaldehyde       WS   Stockfelt et al. 2012 2        
3 266980  Volume Benzene            WS   Stockfelt et al. 2012 1        
4 266980  Volume Benzene            WS   Stockfelt et al. 2012 2        
5 266980  Volume Benzo(GHI)perylene WS   Stockfelt et al. 2012 1        
6 266980  Volume Benzo(GHI)perylene WS   Stockfelt et al. 2012 2        
  Chemical_Class DTXSID        Value
1 VOC            DTXSID5039224 71000
2 VOC            DTXSID5039224 37000
3 VOC            DTXSID3039242 33000
4 VOC            DTXSID3039242 21000
5 PAH            DTXSID5023908 41   
6 PAH            DTXSID5023908 11

In [10]:
dim(combined_df)
dim(filter1_df)

[1] 6581    9

[1] 4189    9

Started with 6581 records, 1363 were removed, leaving 5145. 

Now that each sample type (WS of WF) has at least one experimental value, we'll see if there are at least 50% of data between the sample types.

In [11]:
variable_presence_df = filter1_df %>%
    # if the value isn't MAR count it as being present
    mutate(count = ifelse(Value != 'NA', 1, 0)) %>%
    group_by(Metric, Name) %>%
    # calculating the percentage of variables with data overall
    reframe(Variable_Presence_Percentage = (sum(count)/n()) * 100) %>%
    arrange(-Variable_Presence_Percentage)

# viewing data that passed the filter
keep_variables_df = variable_presence_df %>%
     filter(Variable_Presence_Percentage >= 50) %>%
     unique()

head(keep_variables_df)

Metric Name                       Variable_Presence_Percentage
1 Volume 1,2,4-Trimethylbenzene     100                         
2 Volume 1,3,5-Trimethylbenzene     100                         
3 Volume 1,3-Dihydroxynaphthalene   100                         
4 Volume 1,4-Dichloro-2-butene, cis 100                         
5 Volume 1,4-Dichlorobenzene        100                         
6 Volume 1,8-Dihydroxynaphthalene   100

In [12]:
# only keeping records that passed the background filter
filter2_df = inner_join(keep_variables_df[,1:2], filter1_df) %>%
    # metals don't have a DTXSID, so making that col their name
    mutate(DTXSID = ifelse(DTXSID != 'NA', DTXSID, Name)) 
    

head(filter2_df)

Joining with `by = join_by(Metric, Name)`


Metric Name                       HAWC_ID Data Study                  
1 Volume 1,2,4-Trimethylbenzene     1289926 WF   Wang et al. 2024       
2 Volume 1,2,4-Trimethylbenzene     1306371 WF   Ketcherside et al. 2024
3 Volume 1,3,5-Trimethylbenzene     1289926 WF   Wang et al. 2024       
4 Volume 1,3,5-Trimethylbenzene     1306371 WF   Ketcherside et al. 2024
5 Volume 1,3-Dihydroxynaphthalene   1289821 WF   Liang et al. 2021      
6 Volume 1,4-Dichloro-2-butene, cis 1289926 WF   Wang et al. 2024       
  Replicate Chemical_Class DTXSID         Value
1 1         VOC            DTXSID6021402  730  
2 1         VOC            DTXSID6021402  130  
3 1         VOC            DTXSID6026797  330  
4 1         VOC            DTXSID6026797  110  
5 1         PAH            DTXSID40456587 6    
6 1         VOC            DTXSID3027405  230

In [13]:
dim(filter2_df)

[1] 4189    9

An additional 408 records were removed, leaving 4882.

# Second Variable Background Filter

In the intial filter, samples were combined. However, this time variables (ie. metal or chemical) will be split based on their `Metric` (ie. volume or weight) and then retained if that variable in found in both woodsmoke and wildfire samples within volume or weight samples.

In [14]:
split_filtered_df = filter2_df %>%
    group_by(Data, Metric) %>%
    group_split

split_ws_vol_df = split_filtered_df[[3]]
split_ws_weight_df = split_filtered_df[[4]]
split_wf_vol_df = split_filtered_df[[1]]
split_wf_weight_df = split_filtered_df[[2]]

In [15]:
# seeing how many unique variables are in each df and if they're consistent in each file
length(unique(split_ws_vol_df$Name))
length(unique(split_wf_vol_df$Name))
length(unique(split_ws_weight_df$Name))
length(unique(split_wf_weight_df$Name))

[1] 98

[1] 68

[1] 94

[1] 70

In [16]:
# they're not so first getting variables that are in weight or volume samples
consistent_wf_vol_df = split_wf_vol_df %>%
    filter(Name %in% unique(split_ws_vol_df$Name))
consistent_ws_vol_df = split_ws_vol_df %>%
    filter(Name %in% consistent_wf_vol_df$Name)
consistent_wf_weight_df = split_wf_weight_df %>%
    filter(Name %in% unique(split_ws_weight_df$Name))
consistent_ws_weight_df = split_ws_weight_df %>%
    filter(Name %in% consistent_wf_weight_df$Name)

length(unique(consistent_wf_vol_df$Name))
length(unique(consistent_ws_vol_df$Name))
length(unique(consistent_wf_weight_df$Name))
length(unique(consistent_ws_weight_df$Name))

[1] 33

[1] 33

[1] 34

[1] 34

There were 80, 47, 73 and 61 woodsmoke weight, woodsmoke volume, wildfire weight, and wildfire volume samples, respectively. 26 variables were common between volume samples and 30 were common between weight records and will be retained.

In [17]:
# recombining data
vol_df = rbind(consistent_ws_vol_df, consistent_wf_vol_df)
weight_df = rbind(consistent_ws_weight_df, consistent_wf_weight_df)

head(vol_df)

Metric Name           HAWC_ID Data Study              Replicate
1 Volume Acenaphthylene  429445 WS   Rajput 2010        1        
2 Volume Acenaphthylene 1263480 WS   Burnet et al. 1990 1        
3 Volume Acenaphthylene 1263480 WS   Burnet et al. 1990 2        
4 Volume Acenaphthylene 1263480 WS   Burnet et al. 1990 3        
5 Volume Acenaphthylene 1263484 WS   Leese et al. 1989  1        
6 Volume Acenaphthylene 1263484 WS   Leese et al. 1989  2        
  Chemical_Class DTXSID        Value   
1 PAH            DTXSID3023845 53500   
2 PAH            DTXSID3023845 6187000 
3 PAH            DTXSID3023845 18890500
4 PAH            DTXSID3023845 7806000 
5 PAH            DTXSID3023845 1100000 
6 PAH            DTXSID3023845 2800000

# QRILC Imputation

Imputing non-detect data using QRILC didn't work, so I'm just replacing the non-detect data with zeros.

In [18]:
# determining number of ND and NA values in each df
vol_df %>%
    filter(Value %in% c('NA', 'ND')) %>%
    group_by(Value) %>%
    summarize(n())

weight_df %>%
    filter(Value %in% c('NA', 'ND')) %>%
    group_by(Value) %>%
    summarize(n())

Value n()

Value n()
1 NA     2 
2 ND    34

In [19]:
dim(vol_df)
dim(weight_df)

[1] 744   9

[1] 1411    9

Only the weight dataframe had non-detect values representing MNAR data. The remaining MAR data from the weight df will be imputed using random forest (RF).

In [20]:
mar_weight_df = weight_df %>%
    filter(Value == 'NA')

preimputed_df = anti_join(weight_df, mar_weight_df) 
preimputed_df$Value = as.numeric(preimputed_df$Value)
mar_weight_df$Value = as.numeric(mar_weight_df$Value)

# reordering
preimputed_df = preimputed_df[,c(3,1,4:8,2,9)]

head(preimputed_df)

Joining with `by = join_by(Metric, Name, HAWC_ID, Data, Study, Replicate,
Chemical_Class, DTXSID, Value)`
Warning message:
“NAs introduced by coercion”
Warning message:
“NAs introduced by coercion”


HAWC_ID Metric Data Study                  Replicate Chemical_Class
1 429188  Weight WS   Kasurinen et al. 2018  1         PAH           
2 429188  Weight WS   Kasurinen et al. 2018  2         PAH           
3 429188  Weight WS   Kasurinen et al. 2018  3         PAH           
4 429188  Weight WS   Kasurinen et al. 2018  4         PAH           
5 821855  Weight WS   Erlandsson et al. 2020 1         PAH           
6 821855  Weight WS   Erlandsson et al. 2020 2         PAH           
  DTXSID        Name           Value
1 DTXSID0025654 1-Methylpyrene  0.8 
2 DTXSID0025654 1-Methylpyrene  2.9 
3 DTXSID0025654 1-Methylpyrene  9.1 
4 DTXSID0025654 1-Methylpyrene   NA 
5 DTXSID0025654 1-Methylpyrene 12.0 
6 DTXSID0025654 1-Methylpyrene 19.1

Gave up on QRILC imputation for non-detect values, since I didn't have enough instances to impute within each study. I also couldn't impute across the entire dataset given that number of data instances for each study wasn't the same. REWORD!!!

In [61]:
mnar_df = preimputed_df %>%
    filter(is.na(Value)) 

mnar_df[c(8,9,14:20,23:25,28,34),9] = 0
head(mnar_df)

HAWC_ID Metric Data Study                 Replicate Chemical_Class
1 429188  Weight WS   Kasurinen et al. 2018 4         PAH           
2 271638  Weight WS   Jalava et al. 2012    4         PAH           
3 271638  Weight WS   Jalava et al. 2012    6         PAH           
4 271638  Weight WS   Jalava et al. 2012    7         PAH           
5 271769  Weight WS   Uski et al. 2012      4         PAH           
6 271769  Weight WS   Uski et al. 2012      6         PAH           
  DTXSID        Name           Value
1 DTXSID0025654 1-Methylpyrene NA   
2 DTXSID3021774 Acenaphthene   NA   
3 DTXSID3021774 Acenaphthene   NA   
4 DTXSID3021774 Acenaphthene   NA   
5 DTXSID3021774 Acenaphthene   NA   
6 DTXSID3021774 Acenaphthene   NA

In [47]:
QRILC_imputation = function(dataset){
    set.seed(12)
      # """
    # Creating a quantile normalization function to normalize each sample.
    # :param (input): exposed and unexposed (vehicle) dfs
    # :output: 1 quantile normalized df
    # """
    #ADD WORDS
    wider_dataset = dataset %>% 
        # removing these cols temporarily 
        select(-c("Chemical_Class", "Name")) %>%
        pivot_wider(names_from = DTXSID, values_from = Value)
    
    index_of_last_variable = length(colnames(wider_dataset))

    # normalizing data since that what the QRILC function wants
    # had to pseudo log transform to prevent Inf values
    QRILC_prep = wider_dataset[,6:dim(wider_dataset)[2]] %>%
         mutate_all(., function(x) log10(x + 1)) %>%
         as.matrix()
                    
    imputed_QRILC_object = impute.QRILC(QRILC_prep, tune.sigma = 0.1)
    QRILC_log10_df = data.frame(imputed_QRILC_object[1]) 
    
    # converting back the original scale
    QRILC_df = QRILC_log10_df %>%
        mutate_all(., function(x) 10^x - 1)
    
    imputed_dataset = data.frame(cbind(unique(dataset[,1:5]), QRILC_df)) %>%
         pivot_longer(cols = 6:dim(wider_dataset)[2], names_to = "DTXSID", values_to = "Value")         

    return(imputed_dataset)

}

In [78]:
# will impute within each study
df_for_ids = mnar_df[is.na(mnar_df$Value),]
study_ids = unique(df_for_ids$HAWC_ID)

imputed_df = data.frame()
for (i in 1:length(study_ids)){

    filtered_df = preimputed_df %>%
        filter(HAWC_ID == study_ids[i])

    # calling fn
    imp_df = QRILC_imputation(filtered_df)
    imputed_df = rbind(imputed_df, imp_df)
    
}
# adding back in additional cols
imputed_df = inner_join(unique(imputed_df), unique(weight_df[,1:8]))
imputed_df = imputed_df[,c(1:5,9,6,8,7)]

head(imputed_df)

Joining with `by = join_by(HAWC_ID, Metric, Data, Study, Replicate, DTXSID)`


HAWC_ID Metric Data Study                 Replicate Chemical_Class
1 429188  Weight WS   Kasurinen et al. 2018 1         PAH           
2 429188  Weight WS   Kasurinen et al. 2018 1         PAH           
3 429188  Weight WS   Kasurinen et al. 2018 1         PAH           
4 429188  Weight WS   Kasurinen et al. 2018 1         PAH           
5 429188  Weight WS   Kasurinen et al. 2018 1         PAH           
6 429188  Weight WS   Kasurinen et al. 2018 1         PAH           
  DTXSID        Name                 Value
1 DTXSID0025654 1-Methylpyrene         0.8
2 DTXSID4075459 Benzo(c)phenanthrene   5.5
3 DTXSID8052691 Benzo(j)fluoranthene   3.8
4 DTXSID6062591 Cyclopenta(cd)pyrene   2.0
5 DTXSID3024104 Fluoranthene          75.9
6 DTXSID6024254 Phenanthrene         156.9

In [87]:
imputed_weight_df = rbind(imputed_df, 
                          # adding back in non-imputed data
                          drop_na(preimputed_df) %>%
                              filter(HAWC_ID %notin% unique(mnar_df$HAWC_ID)),
                          # adding back in mar data
                          mar_weight_df, 
                          # adding back in mnar data that was converted to zeros
                          mnar_df[!is.na(mnar_df$Value),])

head(imputed_weight_df)
dim(weight_df)
dim(imputed_weight_df)

HAWC_ID Metric Data Study                 Replicate Chemical_Class
1 429188  Weight WS   Kasurinen et al. 2018 1         PAH           
2 429188  Weight WS   Kasurinen et al. 2018 1         PAH           
3 429188  Weight WS   Kasurinen et al. 2018 1         PAH           
4 429188  Weight WS   Kasurinen et al. 2018 1         PAH           
5 429188  Weight WS   Kasurinen et al. 2018 1         PAH           
6 429188  Weight WS   Kasurinen et al. 2018 1         PAH           
  DTXSID        Name                 Value
1 DTXSID0025654 1-Methylpyrene         0.8
2 DTXSID4075459 Benzo(c)phenanthrene   5.5
3 DTXSID8052691 Benzo(j)fluoranthene   3.8
4 DTXSID6062591 Cyclopenta(cd)pyrene   2.0
5 DTXSID3024104 Fluoranthene          75.9
6 DTXSID6024254 Phenanthrene         156.9

[1] 1411    9

[1] 1294    9

In [89]:
dim(imputed_df)
dim(mar_weight_df)
dim(mnar_df[!is.na(mnar_df$Value),])
dim(drop_na(preimputed_df) %>%
                              filter(HAWC_ID %notin% unique(mnar_df$HAWC_ID)))

[1] 335   9

[1] 2 9

[1] 14  9

[1] 943   9

In [25]:
preimputed_df %>%
    # changing mnar data to zeros
    mutate(Value = ifelse(is.na(Value), 0, Value)) %>%
    # adding back in mar data
    full_join(mar_weight_df)

Joining with `by = join_by(HAWC_ID, Metric, Data, Study, Replicate,
Chemical_Class, DTXSID, Name, Value)`


HAWC_ID Metric Data Study                    Replicate Chemical_Class
1     429188 Weight WS   Kasurinen et al. 2018    1         PAH           
2     429188 Weight WS   Kasurinen et al. 2018    2         PAH           
3     429188 Weight WS   Kasurinen et al. 2018    3         PAH           
4     429188 Weight WS   Kasurinen et al. 2018    4         PAH           
5     821855 Weight WS   Erlandsson et al. 2020   1         PAH           
6     821855 Weight WS   Erlandsson et al. 2020   2         PAH           
7    1257056 Weight WS   McCarrick et al. 2024    1         PAH           
8     821855 Weight WS   Erlandsson et al. 2020   1         PAH           
9     821855 Weight WS   Erlandsson et al. 2020   2         PAH           
10   1257056 Weight WS   McCarrick et al. 2024    1         PAH           
11    821855 Weight WS   Erlandsson et al. 2020   1         PAH           
12    821855 Weight WS   Erlandsson et al. 2020   2         PAH           
13   1257056 Weight WS   McCarrick et al. 2024    1         PAH           
14    267105 Weight WS   Danielsen et al. 2011    1         PAH           
15    267105 Weight WS   Danielsen et al. 2011    2         PAH           
16    279652 Weight WS   Sussan et al. 2014       1         PAH           
17    821855 Weight WS   Erlandsson et al. 2020   1         PAH           
18    821855 Weight WS   Erlandsson et al. 2020   2         PAH           
19    271638 Weight WS   Jalava et al. 2012       1         PAH           
20    271638 Weight WS   Jalava et al. 2012       2         PAH           
21    271638 Weight WS   Jalava et al. 2012       3         PAH           
22    271638 Weight WS   Jalava et al. 2012       4         PAH           
23    271638 Weight WS   Jalava et al. 2012       5         PAH           
24    271638 Weight WS   Jalava et al. 2012       6         PAH           
25    271638 Weight WS   Jalava et al. 2012       7         PAH           
26    271769 Weight WS   Uski et al. 2012         1         PAH           
27    271769 Weight WS   Uski et al. 2012         2         PAH           
28    271769 Weight WS   Uski et al. 2012         3         PAH           
29    271769 Weight WS   Uski et al. 2012         4         PAH           
30    271769 Weight WS   Uski et al. 2012         5         PAH           
⋮    ⋮       ⋮      ⋮    ⋮                        ⋮         ⋮             
1382  429319 Weight WF   Kim et al. 2018          8         Metal         
1383  981013 Weight WF   Jalava et al. 2006       1         Metal         
1384  981013 Weight WF   Jalava et al. 2006       2         Metal         
1385  981013 Weight WF   Jalava et al. 2006       3         Metal         
1386  981013 Weight WF   Jalava et al. 2006       4         Metal         
1387  981110 Weight WF   Verma et al. 2009        1         Metal         
1388 1104106 Weight WF   Alexakis 2020            1         Metal         
1389 1104106 Weight WF   Alexakis 2020            2         Metal         
1390 1257076 Weight WF   Villaruel et al. 2024    1         Metal         
1391 1285773 Weight WF   Burton et al. 2016       1         Metal         
1392 1285774 Weight WF   Santin et al. 2015       1         Metal         
1393 1285775 Weight WF   Plumlee et al. 2007      1         Metal         
1394 1285872 Weight WF   Correa Velez et al. 2024 1         Metal         
1395 1285872 Weight WF   Correa Velez et al. 2024 2         Metal         
1396 1285970 Weight WF   Sanchez et al. 2023      1         Metal         
1397 1285992 Weight WF   Holder et al. 2023       1         Metal         
1398 1285992 Weight WF   Holder et al. 2023       2         Metal         
1399 1285992 Weight WF   Holder et al. 2023       3         Metal         
1400 1285992 Weight WF   Holder et al. 2023       4         Metal         
1401 1289746 Weight WF   Das et al. 2019          1         Metal         
1402 1289764 Weight WF   Gabet and Bookter 2011   1         Metal         
1403 1289764 Weight WF   Gabet

In [20]:
# avg_df = preimputed_df %>%
#     group_by(HAWC_ID, Name) %>%
#     mutate(Avg_Value = mean(Value)) %>%
#     select(-c("Value", "Replicate")) %>%
#     unique()

# head(avg_df)

# RF Imputation

In [23]:
preimputed_weight_df = weight_df %>%
    mutate(Value = ifelse(Value == 'ND', '0', Value)) %>%
    group_by(Data, HAWC_ID, Name) %>%
    mutate(Avg_Value = mean(as.numeric(Value, na.rm = TRUE))) %>%
    select(-c("Replicate", "Value")) %>%
    unique()
    # creating sample ids
    #unite("SampleID", c(colnames(weight_df)[c(3,4,6)]), sep = "_")

#preimputed_weight_df$Value = as.numeric(preimputed_weight_df$Value)

head(preimputed_weight_df)

Warning message:
“There were 220 warnings in `mutate()`.
The first warning was:
ℹ In argument: `Avg_Value = mean(as.numeric(Value, na.rm = TRUE))`.
ℹ In group 244: `Data = "WS"`, `HAWC_ID = 267105`, `Name = "Acenaphthene"`.
Caused by warning in `mean()`:
! NAs introduced by coercion
ℹ Run `dplyr::last_dplyr_warnings()` to see the 219 remaining warnings.”


Metric Name                    Data HAWC_ID Study                 
1 Weight 2,3-Dimethylnaphthalene WS    821855 Erlandsson et al. 2020
2 Weight 2,3-Dimethylnaphthalene WS   1257056 McCarrick et al. 2024 
3 Weight 2-Methylchrysene        WS    821855 Erlandsson et al. 2020
4 Weight 2-Methylchrysene        WS   1257056 McCarrick et al. 2024 
5 Weight 2-Methylphenanthrene    WS    267105 Danielsen et al. 2011 
6 Weight 2-Methylphenanthrene    WS    821855 Erlandsson et al. 2020
  Chemical_Class DTXSID        Avg_Value
1 PAH            DTXSID2060383  0.05    
2 PAH            DTXSID2060383  0.04    
3 PAH            DTXSID8074819  9.75    
4 PAH            DTXSID8074819 15.45    
5 PAH            DTXSID4020878 23.50    
6 PAH            DTXSID4020878  3.50

In [31]:
ncol(preimputed_weight_df)

[1] 8

In [58]:
random_forest_imputation = function(dataset){
    # ADD WORDS
    wider_dataset = dataset %>%
        pivot_wider(names_from = DTXSID, values_from = Avg_Value)
    preimputed_df = wider_dataset[,8:ncol(wider_dataset)]
    # had to drop rows (studies) that only had missing data
    df = preimputed_df[rowSums(is.na(preimputed_df)) != ncol(preimputed_df), ]
    
    imputed_RF_object = missForest(as.matrix(df))
    imputed_RF_df = imputed_RF_object$ximp
    imputed_final_df = data.frame(cbind(wider_dataset[,1:7], imputed_RF_df)) %>%
        pivot_longer(cols = 8:ncol(wider_dataset), names_to = "DTXSID", values_to = "Value")
    
    return(imputed_final_df)
}


# calling fn
a = random_forest_imputation(preimputed_weight_df %>% filter(Data == 'WS'))

Warning message in randomForest.default(x = obsX, y = obsY, ntree = ntree, mtry = mtry, :
“The response has five or fewer unique values.  Are you sure you want to do regression?”
Warning message in randomForest.default(x = obsX, y = obsY, ntree = ntree, mtry = mtry, :
“The response has five or fewer unique values.  Are you sure you want to do regression?”
Warning message in randomForest.default(x = obsX, y = obsY, ntree = ntree, mtry = mtry, :
“The response has five or fewer unique values.  Are you sure you want to do regression?”
Warning message in randomForest.default(x = obsX, y = obsY, ntree = ntree, mtry = mtry, :
“The response has five or fewer unique values.  Are you sure you want to do regression?”
Warning message in randomForest.default(x = obsX, y = obsY, ntree = ntree, mtry = mtry, :
“The response has five or fewer unique values.  Are you sure you want to do regression?”
Warning message in randomForest.default(x = obsX, y = obsY, ntree = ntree, mtry = mtry, :
“The response h

ERROR: Error in if (n == 0) stop("data (x) has 0 rows"): argument is of length zero


In [48]:
l = data.frame(a[,8:37])
dim(l)

[1] 698  30

In [49]:
l[rowSums(is.na(l)) != ncol(l), ]

DTXSID8074819 DTXSID4020878 B   DTXSID4075459 DTXSID4020161 Cu       
3    9.75             NA         NA        NA        NA                NA
4   15.45             NA         NA        NA        NA                NA
5      NA         23.500         NA        NA        NA                NA
6      NA          3.500         NA        NA        NA                NA
7      NA          5.661         NA        NA        NA                NA
8      NA             NA        288        NA        NA                NA
9      NA             NA         NA  21.00000        NA                NA
10     NA             NA         NA 235.42857        NA                NA
11     NA             NA         NA   0.45000        NA                NA
12     NA             NA         NA  13.10000        NA                NA
13     NA             NA         NA 257.95000        NA                NA
14     NA             NA         NA   0.00193        NA                NA
15     NA             NA         NA  14.98500        NA                NA
16     NA             NA         NA 234.48571        NA                NA
17     NA             NA         NA        NA     0.045                NA
18     NA             NA         NA        NA     0.080                NA
19     NA             NA         NA        NA        NA          187.0000
20     NA             NA         NA        NA        NA          375.0000
21     NA             NA         NA        NA        NA           50.0000
22     NA             NA         NA        NA        NA            4.1000
23     NA             NA         NA        NA        NA           28.8000
24     NA             NA         NA        NA        NA           28.7275
25     NA             NA         NA        NA        NA          278.5000
26     NA             NA         NA        NA        NA          151.3333
27     NA             NA         NA        NA        NA         4100.0000
28     NA             NA         NA        NA        NA            4.1000
29     NA             NA         NA        NA        NA            0.3500
30     NA             NA         NA        NA        NA          266.0000
31     NA             NA         NA        NA        NA          233.0000
32     NA             NA         NA        NA        NA           70.0000
⋮   ⋮             ⋮             ⋮   ⋮             ⋮             ⋮        
669 NA            NA            NA  NA            NA            NA       
670 NA            NA            NA  NA            NA            NA       
671 NA            NA            NA  NA            NA            NA       
672 NA            NA            NA  NA            NA            NA       
673 NA            NA            NA  NA            NA            NA       
674 NA            NA            NA  NA            NA            NA       
675 NA            NA            NA  NA            NA            NA       
676 NA            NA            NA  NA            NA            NA       
677 NA            NA            NA  NA            NA            NA       
678 NA            NA            NA  NA            NA            NA       
679 NA            NA            NA  NA            NA            NA       
680 NA            NA            NA  NA            NA            NA       
681 NA            NA            NA  NA            NA            NA       
682 NA            NA            NA  NA            NA            NA       
683 NA            NA            NA  NA            NA            NA       
684 NA            NA            NA  NA            NA            NA       
685 NA            NA            NA  NA            NA            NA       
686 NA            NA            NA  NA            NA            NA       
687 NA            NA            NA  NA            NA            NA       
688 NA            NA            NA  NA            NA            NA       
689 NA            NA            NA  NA            NA            NA       
690 NA            NA            NA  NA            NA            NA       
691 NA  

In [55]:
l

DTXSID8074819 DTXSID4020878 B   DTXSID4075459 DTXSID4020161 Cu       
1      NA             NA         NA        NA        NA                NA
2      NA             NA         NA        NA        NA                NA
3    9.75             NA         NA        NA        NA                NA
4   15.45             NA         NA        NA        NA                NA
5      NA         23.500         NA        NA        NA                NA
6      NA          3.500         NA        NA        NA                NA
7      NA          5.661         NA        NA        NA                NA
8      NA             NA        288        NA        NA                NA
9      NA             NA         NA  21.00000        NA                NA
10     NA             NA         NA 235.42857        NA                NA
11     NA             NA         NA   0.45000        NA                NA
12     NA             NA         NA  13.10000        NA                NA
13     NA             NA         NA 257.95000        NA                NA
14     NA             NA         NA   0.00193        NA                NA
15     NA             NA         NA  14.98500        NA                NA
16     NA             NA         NA 234.48571        NA                NA
17     NA             NA         NA        NA     0.045                NA
18     NA             NA         NA        NA     0.080                NA
19     NA             NA         NA        NA        NA          187.0000
20     NA             NA         NA        NA        NA          375.0000
21     NA             NA         NA        NA        NA           50.0000
22     NA             NA         NA        NA        NA            4.1000
23     NA             NA         NA        NA        NA           28.8000
24     NA             NA         NA        NA        NA           28.7275
25     NA             NA         NA        NA        NA          278.5000
26     NA             NA         NA        NA        NA          151.3333
27     NA             NA         NA        NA        NA         4100.0000
28     NA             NA         NA        NA        NA            4.1000
29     NA             NA         NA        NA        NA            0.3500
30     NA             NA         NA        NA        NA          266.0000
⋮   ⋮             ⋮             ⋮   ⋮             ⋮             ⋮        
669 NA            NA            NA  NA            NA            NA       
670 NA            NA            NA  NA            NA            NA       
671 NA            NA            NA  NA            NA            NA       
672 NA            NA            NA  NA            NA            NA       
673 NA            NA            NA  NA            NA            NA       
674 NA            NA            NA  NA            NA            NA       
675 NA            NA            NA  NA            NA            NA       
676 NA            NA            NA  NA            NA            NA       
677 NA            NA            NA  NA            NA            NA       
678 NA            NA            NA  NA            NA            NA       
679 NA            NA            NA  NA            NA            NA       
680 NA            NA            NA  NA            NA            NA       
681 NA            NA            NA  NA            NA            NA       
682 NA            NA            NA  NA            NA            NA       
683 NA            NA            NA  NA            NA            NA       
684 NA            NA            NA  NA            NA            NA       
685 NA            NA            NA  NA            NA            NA       
686 NA            NA            NA  NA            NA            NA       
687 NA            NA            NA  NA            NA            NA       
688 NA            NA            NA  NA            NA            NA       
689 NA            NA            NA  NA            NA            NA       
690 NA            NA            NA  NA            NA            NA       
691 NA  

In [20]:
# # exporting
# write.xlsx(processed_df, paste0(Output,"/", "Processed_Data_", cur_date, ".xlsx"), 
#            rowNames = FALSE)